### PR TITLE
add code to replace AddMatMat with AddMatMatBatched in Convolutional1…

### PR DESCRIPTION
**Pull request by freewym**
…dComponent

Hi Dan,

I implemented AddMatBlocks() in cu-matrix.cc and its corresponding custon kernel function, with no support for transpose.
I added a test in cu-matrix-test.cc for testing AddMatBlocks().

valgrind and cuda-memcheck both indicate no leaks on cu-matrix-test and nnet-component.

I left bias_grad.AddRowSumMat() in the for-loop because I am still not clear how to do it in one call given the inconsistent dims between bias_grad and cols of out_deriv. Could you please clarify it?

BTW, there is a member function AddMatBlock() in CuMatrixBase, which is easily confused with  AddMatBlocks(). Would it be better to propose another name for AddMatBlocks?

The github said "Can't automatically merge" with kaldi-asr/kaldi/master. Is it because there exists some line conflicts?

Thanks,
Yiming

### CODE REVIEW

**Comment by naxingyu** [comment]
You may get help from here.
https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/

It usually happen when your commit and another commit are changing the 
same lines of code.

**Comment by freewym** [comment]
@naxingyu Thanks.

**Comment by danpovey** [comment]
Regarding "I leave bias_grad.AddRowSumMat() in the for loop because I am
still not clear how to do it in one call given the inconsistent dims
between bias_grad and cols of out_deriv. Could you please clarify it?"--

I mean you should create a temporary matrix with the same number of rows as
out_deriv, but the correct number of columns (same as dim of bias_grad),
and use AddMatBlocks() to create the temporary matrix.  You'd have to make
AddMatBlocks work if the num-cols differs.  (You could make it work even if
both num-cols and num-rows differ, by having two loops in the code).

"BTW, there is a member function AddMatBlock() in CuMatrixBase, which is
easily confused with AddMatBlocks(). Is it better to propose another name
for AddMatBlocks?"

I know about that... I couldn't think of a better alternative.

Dan

**Comment by freewym** [comment]
I have updated the call of bias_grad.AddRowSumMat() , as long as made AddMatBlock() to support both row blocks and col blocks. These 2 commits have be squashed into a single one.

Yiming

File: src/nnet2/nnet-component.cc [hunk 1]
```diff
@@ -4170,13 +4211,42 @@ void Convolutional1dComponent::Update(const CuMatrixBase<BaseFloat> &in_value,
...
-  for (int32 p = 0; p < num_patches; p++) { // sum
-    CuSubMatrix<BaseFloat> diff_patch(out_deriv.ColRange(p * num_filters,
-                                                         num_filters));
-    CuSubMatrix<BaseFloat> patch(patches.ColRange(p * filter_dim, filter_dim));
-    filters_grad.AddMatMat(1.0, diff_patch, kTrans, patch, kNoTrans, 1.0);
-    bias_grad.AddRowSumMat(1.0, diff_patch, 1.0);
+  
+  //allocate a single large matrix holding the smaller matrices from 
+  //the vector container filters_grad_batch
+  CuMatrix<BaseFloat> *filters_grad_blocks_batch = new CuMatrix<BaseFloat>(
```
Comment (danpovey) [inline src/nnet2/nnet-component.cc]: there no need to allocate this on the heap.

**Comment by danpovey** [comment]
Please make sure your AddMatBatch has a speed-test routine.
Also, let me know when you think it's ready to merge.
Dan

**Comment by danpovey** [comment]
Sorry I mean  AddMatBlocks
Dan

**Comment by freewym** [comment]
Hi Dan,

I have pushed a single commit. Let me know if there are any problems.

According to the speed test, AddMatBlocks improves over AddMat on GPU (I removed the speed test for AddMat in the final commit):

LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 16 numRowBlocks = 4 numColBlocks = 4, speed was 0.000310383 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 32 numRowBlocks = 4 numColBlocks = 4, speed was 0.00175293 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 64 numRowBlocks = 4 numColBlocks = 4, speed was 0.0118271 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 128 numRowBlocks = 4 numColBlocks = 4, speed was 0.0301476 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 256 numRowBlocks = 4 numColBlocks = 4, speed was 0.111097 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 512 numRowBlocks = 4 numColBlocks = 4, speed was 0.438642 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 1024 numRowBlocks = 4 numColBlocks = 4, speed was 1.71881 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 16 numRowBlocks = 4numColBlocks = 4, speed was 0.00687876 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 32 numRowBlocks = 4 numColBlocks = 4, speed was 0.0460843 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 64 numRowBlocks = 4 numColBlocks = 4, speed was 0.186284 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 128 numRowBlocks = 4 numColBlocks = 4, speed was 0.752389 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 256 numRowBlocks = 4 numColBlocks = 4, speed was 2.46113 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 512 numRowBlocks = 4 numColBlocks = 4, speed was 4.26382 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 1024 numRowBlocks = 4 numColBlocks = 4, speed was 7.09727 gigaflops.

Yiming

**Comment by danpovey** [comment]
OK, very good.  But please make sure the TestCuMatrixAddMat() test code
gets made the same as it was before, when you check in.  I want to keep
that as a test for that one routine.
Dan

**Comment by freewym** [comment]
Done.

**Comment by danpovey** [comment]
There are conflicts in the pull request-- can you please attempt to rebase
against master, e.g.

 git rebase upstream/master

and fix any conflicts?  Get someone who knows git to help you.

Dan

**Comment by freewym** [comment]
I figured out: the conflicts lines are in two files.
1. cu-matrix.cc. I removed line 896 and added back at line 931
2. cu-matrix-speed-test.cc: it seems that you added TestCuMatrixResize at line 748/790 after my last checkout. I retained this one.

Ran the test after solving the conflicts. Passed. Checked in the new commit.

**Comment by danpovey** [comment]
You implemented the transposed version of AddMatBLocks, as far as I can
see, but I don't see the test code testing that.
Dan

**Comment by freewym** [comment]
I didn't implement it. The function  _add_mat_blocks_trans is actually identical to the no-trans version _add_mat_blocks in cu-kernels.cu. I did this because I noticed that _add_mat and _add_mat_trans seem identical.

**Comment by danpovey** [comment]
But your code does support it-- it shouldn't silently give wrong results if
it doesn't support something, it should crash with an error message.
But actually, look more carefully at where index_src is declared-- those
functions are different.
It would be better to implement and test it.
Dan

**Comment by freewym** [comment]
Sorry. I will fix/implement that.

**Comment by freewym** [comment]
Please check the new one. The trans version of AddMatBlocks passed the test.

Yiming

**Comment by danpovey** [comment]
Thanks!
Merging.
Dan

**Comment by danpovey** [comment]
@naxingyu, if you have a test setup for the convolutional stuff now might be a good time to re-test.

**Comment by naxingyu** [comment]
OK. Thanks.

**Pull request merged**